File: src/wallet.rs
use serde::{Deserialize, Serialize};
use base58::ToBase58;
use crate::{
    crypto::{generate_mnemonic, mnemonic_to_seed, derive_key_from_seed},
    storage::WalletStorage,
    network::NetworkClient,
    error::WalletError,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Wallet {
    pub name: String,
    pub seed_phrase: String,
    pub addresses: Vec<String>,
    pub created_at: chrono::DateTime<chrono::Utc>,
}

impl Wallet {
    pub fn create_new(name: &str) -> Result<Self, WalletError> {
        let seed_phrase = generate_mnemonic()?;
        let mut wallet = Wallet {
            name: name.to_string(),
            seed_phrase: seed_phrase.clone(),
            addresses: Vec::new(),
            created_at: chrono::Utc::now(),
        };
        
        // Generate initial address
        let address = wallet.generate_address(0)?;
        wallet.addresses.push(address);
        
        // Save wallet
        wallet.save()?;
        
        Ok(wallet)
    }
    
    pub fn from_seed_phrase(seed_phrase: &str, name: &str) -> Result<Self, WalletError> {
        // Validate seed phrase
        let _ = mnemonic_to_seed(seed_phrase)?;
        
        let mut wallet = Wallet {
            name: name.to_string(),
            seed_phrase: seed_phrase.to_string(),
            addresses: Vec::new(),
            created_at: chrono::Utc::now(),
        };
        
        // Generate initial address
        let address = wallet.generate_address(0)?;
        wallet.addresses.push(address);
        
        // Save wallet
        wallet.save()?;
        
        Ok(wallet)
    }
    
    pub fn load(name: &str) -> Result<Self, WalletError> {
        WalletStorage::load(name)
    }
    
    pub fn save(&self) -> Result<(), WalletError> {
        WalletStorage::save(self)
    }
    
    pub fn generate_address(&self, index: u32) -> Result<String, WalletError> {
        let seed = mnemonic_to_seed(&self.seed_phrase)?;
        let derivation_path = format!("m/44'/0'/0'/0/{}", index);
        let _key = derive_key_from_seed(&seed, &derivation_path)?;
        
        // For demo purposes, generate a simple address
        // In a real implementation, you'd use the derived key to create a proper public key
        let mut address_bytes = vec![0x00]; // Mainnet prefix
        address_bytes.extend_from_slice(&seed[..20]); // Use part of seed as address
        address_bytes.extend_from_slice(&[0x00, 0x00, 0x00, 0x00]); // Simple checksum
        
        Ok(address_bytes.to_base58())
    }
    
    pub fn get_addresses(&self) -> &Vec<String> {
        &self.addresses
    }
    
    pub fn add_new_address(&mut self) -> Result<String, WalletError> {
        let index = self.addresses.len() as u32;
        let address = self.generate_address(index)?;
        self.addresses.push(address.clone());
        self.save()?;
        Ok(address)
    }
    
    pub fn get_seed_phrase(&self) -> &str {
        &self.seed_phrase
    }
    
    pub async fn get_balance(&self, network: &str) -> Result<f64, WalletError> {
        let client = NetworkClient::new(network)?;
        let mut total_balance = 0.0;
        
        for address in &self.addresses {
            let balance = client.get_balance(address).await?;
            total_balance += balance;
        }
        
        Ok(total_balance)
    }
    
    pub async fn send_transaction(&mut self, to_address: &str, amount: f64, network: &str) -> Result<String, WalletError> {
        // Validate address format
        if !self.is_valid_address(to_address) {
            return Err(WalletError::InvalidAddress(format!("Invalid recipient address: {}", to_address)));
        }
        
        // Check balance
        let balance = self.get_balance(network).await?;
        if balance < amount {
            return Err(WalletError::InsufficientFunds);
        }
        
        // Create transaction
        let client = NetworkClient::new(network)?;
        let tx_hash = client.send_transaction(&self.addresses[0], to_address, amount).await?;
        
        Ok(tx_hash)
    }
    
    fn is_valid_address(&self, address: &str) -> bool {
        // Basic Bitcoin address validation
        if address.len() < 26 || address.len() > 35 {
            return false;
        }
        
        // Check if it's base58 encoded
        match base58::FromBase58::from_base58(address) {
            Ok(_) => true,
            Err(_) => false,
        }
    }
}

File: backend/src/wallet.cpp
#include "wallet.h"
#include "storage.h"
#include "network.h"
#include "crypto.h"
#include <chrono>
#include <iostream>

namespace crypto_wallet {

Wallet Wallet::create_new(const std::string& name) {
    Wallet wallet;
    wallet.name = name;
    wallet.seed_phrase = Crypto::generate_mnemonic();
    wallet.created_at = std::chrono::system_clock::now();
    
    // Generate initial address
    auto address = wallet.generate_address(0);
    wallet.addresses.push_back(address);
    
    // Save wallet
    wallet.save();
    
    return wallet;
}

Wallet Wallet::from_seed_phrase(const std::string& seed_phrase, const std::string& name) {
    // Validate seed phrase
    try {
        Crypto::mnemonic_to_seed(seed_phrase);
    } catch (...) {
        throw WalletError::invalid_seed_phrase();
    }
    
    Wallet wallet;
    wallet.name = name;
    wallet.seed_phrase = seed_phrase;
    wallet.created_at = std::chrono::system_clock::now();
    
    // Generate initial address
    auto address = wallet.generate_address(0);
    wallet.addresses.push_back(address);
    
    // Save wallet
    wallet.save();
    
    return wallet;
}

Wallet Wallet::load(const std::string& name) {
    return WalletStorage::load(name);
}

void Wallet::save() const {
    WalletStorage::save(*this);
}

std::string Wallet::generate_address(uint32_t index) const {
    auto seed = Crypto::mnemonic_to_seed(seed_phrase);
    std::string derivation_path = "m/44'/0'/0'/0/" + std::to_string(index);
    auto key = Crypto::derive_key_from_seed(seed, derivation_path);
    
    // For demo purposes, generate a simple address
    std::vector<uint8_t> address_bytes = {0x00}; // Mainnet prefix
    address_bytes.insert(address_bytes.end(), key.begin(), key.begin() + 20);
    address_bytes.insert(address_bytes.end(), {0x00, 0x00, 0x00, 0x00}); // Simple checksum
    
    return Crypto::base58_encode(address_bytes);
}

std::string Wallet::add_new_address() {
    uint32_t index = static_cast<uint32_t>(addresses.size());
    auto address = generate_address(index);
    addresses.push_back(address);
    save();
    return address;
}

double Wallet::get_balance(const std::string& network) const {
    auto client = NetworkClient::create(network);
    double total_balance = 0.0;
    
    for (const auto& address : addresses) {
        try {
            auto balance = client->get_balance(address);
            total_balance += balance;
        } catch (const WalletError& e) {
            // Log error but continue with other addresses
            std::cerr << "Error getting balance for address " << address << ": " << e.what() << std::endl;
        }
    }
    
    return total_balance;
}

std::string Wallet::send_transaction(
    const std::string& to_address, 
    double amount, 
    const std::string& network
) const {
    // Validate address format
    if (!is_valid_address(to_address)) {
        throw WalletError::invalid_address("Invalid recipient address: " + to_address);
    }
    
    // Check balance
    auto balance = get_balance(network);
    if (balance < amount) {
        throw WalletError::insufficient_funds();
    }
    
    // Create transaction
    auto client = NetworkClient::create(network);
    auto tx_hash = client->send_transaction(addresses[0], to_address, amount);
    
    return tx_hash;
}

bool Wallet::is_valid_address(const std::string& address) const {
    return Crypto::is_valid_address(address);
}

} // namespace crypto_wallet
